src/model_selection/mod.rs

```diff
@@ -81,6 +81,100 @@ pub fn train_test_split<T: RealNumber, M: Matrix<T>>(
     (x_train, x_test, y_train, y_test)
 }
 
+///
+/// KFold Cross-Validation
+///
+trait BaseKFold {
+    /// Returns integer indices corresponding to test sets
+    fn test_indices<T: RealNumber, M: Matrix<T>>(&self, x: &M) -> Vec<Vec<usize>>;
+
+    /// Returns masksk corresponding to test sets
+    fn test_masks<T: RealNumber, M: Matrix<T>>(&self, x: &M) -> Vec<Vec<bool>>;
+
+    /// Return a tuple containing the the training set indices for that split and
+    /// the testing set indices for that split.
+    fn split<T: RealNumber, M: Matrix<T>>(&self, x: &M) -> Vec<(Vec<usize>, Vec<usize>)>;
+}
+
+///
+/// An implementation of KFold
+///
+pub struct KFold {
+    n_splits: i32, // cannot exceed std::usize::MAX
+                   // TODO: to be implemented later
+                   // shuffle: bool,
+                   // random_state: i32,
+}
+
+impl BaseKFold for KFold {
+    fn test_indices<T: RealNumber, M: Matrix<T>>(&self, x: &M) -> Vec<Vec<usize>> {
+        // number of samples (rows) in the matrix
+        let n_samples: usize = x.shape().1;
+
+        // initialise indices
+        let indices: Vec<usize> = (0..n_samples).collect();
+
+        //  return a new array of given shape n_split, filled with each element of n_samples divided by n_splits.
+        let mut fold_sizes = vec![n_samples / self.n_splits as usize; self.n_splits as usize];
+
+        // increment by one if odd
+        for i in 0..(n_samples % self.n_splits as usize) {
+            fold_sizes[i] = fold_sizes[i] + 1;
+        }
+
+        // generate the right array of arrays for test indices
+        let mut return_values: Vec<Vec<usize>> = Vec::with_capacity(self.n_splits as usize);
+        let mut current: usize = 0;
+        for fold_size in fold_sizes.drain(..) {
+            let stop = current + fold_size;
+            return_values.push(indices[current..stop].to_vec());
+            current = stop
+        }
+
+        return return_values;
+    }
+
+    fn test_masks<T: RealNumber, M: Matrix<T>>(&self, x: &M) -> Vec<Vec<bool>> {
+        let mut return_values: Vec<Vec<bool>> = Vec::with_capacity(self.n_splits as usize);
+        for test_index in self.test_indices(x).drain(..) {
+            // init mask
+            let mut test_mask = vec![false; x.shape().0];
+            // set mask's indices to true according to test indices
+            for i in test_index {
+                test_mask[i] = true; // can be implemented with map()
+            }
+            return_values.push(test_mask);
+        }
+        return return_values;
+    }
+
+    fn split<T: RealNumber, M: Matrix<T>>(&self, x: &M) -> Vec<(Vec<usize>, Vec<usize>)> {
```

Comment:
Looks like this method does not split **x** in the same way as Scikit  does. For example, 3-fold split of 3x4 matrix in Python would look like this:
![image](https://user-images.githubusercontent.com/1102497/95691412-dd13cf00-0bd3-11eb-86c2-634b0bf29456.png). 

While our implementation outputs this:
![image](https://user-images.githubusercontent.com/1102497/95691482-56abbd00-0bd4-11eb-89c1-6e56662df0d2.png)
![image](https://user-images.githubusercontent.com/1102497/95691491-5dd2cb00-0bd4-11eb-9453-1e192c129753.png)

I suggest this test to verify parity between Scikit learn implementation and ours:
```suggestion
#[test]
    fn numpy_parity_test() {
        let k = KFold { n_splits: 3 };
        let x: DenseMatrix<f64> = DenseMatrix::rand(10, 4);
        let expected: Vec<(Vec<usize>, Vec<usize>)> = vec!(
            (vec!(4, 5, 6, 7, 8, 9), vec!(0, 1, 2, 3)),
            (vec!(0, 1, 2, 3, 7, 8, 9), vec!(4, 5, 6)),
            (vec!(0, 1, 2, 3, 4, 5, 6), vec!(7, 8, 9))
        );
        for ((train, test), (expected_train, expected_test)) in k.split(&x).into_iter().zip(expected) {
            assert_eq!(test, expected_test);
            assert_eq!(train, expected_train);
        }        
    }
```